### PR TITLE
New Histograms [WIP]

### PR DESCRIPTION
**Pull request by laullon**
**Description**
Change the histogram logic to discard samples older than 1 minute before doing the calculation, and auto clean old samples every 5 seconds.

**Todo-List**
* [ ] Add docs
* [ ] Cleanup code
* [ ] Make it user configurable
* [ ] Test it with the latest @venilnoronha changes
* [ ] Test it on GCP

Solve issue https://github.com/vmware/wavefront-adapter-for-istio/issues/13

### CODE REVIEW

File: wavefront/timeUniformSample.go [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+	"sync"
+	"time"
+
+	metrics "github.com/rcrowley/go-metrics"
+)
+
+type TimeUniformSample struct {
+	mutex    sync.Mutex
+	values   Queue
+	lifeTime time.Duration
```
Comment (venilnoronha) [inline wavefront/timeUniformSample.go]: nit: `lifeTime` -> `lifetime` here and at other places

File: wavefront/timeUniformSample.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+}
+
+var sampleList = make([]*TimeUniformSample, 0)
+var ticker *time.Ticker
+
+func NewTimeUniformSample(lifeTime time.Duration, recycleTime time.Duration) metrics.Sample {
+	sample := &TimeUniformSample{lifeTime: lifeTime, values: Queue{}}
+
+	if ticker == nil {
+		ticker = time.NewTicker(recycleTime)
```
Comment (venilnoronha) [inline wavefront/timeUniformSample.go]: What if different `TimeUniformSample` instances have different `recycleTime`?

File: wavefront/timeUniformSample.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+}
+
+// Variance returns the variance of the values in the sample.
+func (s *TimeUniformSample) Variance() float64 {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return metrics.SampleVariance(s.rawValues())
+}
+
+type SampleValue struct {
```
Comment (venilnoronha) [inline wavefront/timeUniformSample.go]: nit: `SampleValue` -> `sampleValue` as it's not being used outside this code.

File: wavefront/timeUniformSample.go [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+	return result, true
+}
+
+func (q *Queue) Peek() (interface{}, bool) {
+	if q.len <= 0 {
+		return nil, true
+	}
+	result := q.content[q.readHead]
+	return result, false
+}
```
Comment (venilnoronha) [inline wavefront/timeUniformSample.go]: Do we need the whole `Queue` structure here? I think a Go slice could do the job.

https://github.com/golang/go/wiki/SliceTricks
https://stackoverflow.com/a/26863706

**Comment by venilnoronha** [comment]
Ping CI

**Comment by venilnoronha** [comment]
Please rebase with master to fix the build error.

File: wavefront/timeUniformSample.go (hunk 4 above)
Comment (laullon) [reply wavefront/timeUniformSample.go]: Queue structure is mostly for code readability...

Also, I'm not a Go expert, but `queue = queue[1:]` looks bad to me...
A comment on your stackoverflow link:

```
This is not a good approach, as every time queue[1:] is done, it moves the start of the slice to point to the next element, but does not release the spaced used by the dequeued element. Effectively it has unbounded memory growth in the slice storage. Additionally, if the queued element is a pointer, or a struct containing pointers, then the underlying slice storage will retain the memory to the dequeued element, causing a further memory leak. – tul Jun 8 at 10:11
```

but now we have memory metrics, so will try

Comment (venilnoronha) [reply wavefront/timeUniformSample.go]: Let's retain the `Queue` struct then.

File: wavefront/timeUniformSample.go [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+		sample, empty := s.values.Peek()
+		if empty {
+			needPop = false
+		} else {
+			needPop = (now.Sub(sample.(*sampleValue).time).Seconds() > s.lifetime.Seconds())
+			if needPop {
+				s.values.Pop()
+			}
+		}
+	}
```
Comment (venilnoronha) [inline wavefront/timeUniformSample.go]: This block could be simplified, like so:

```go
for {
	if sample, empty := s.values.Peek(); empty {
		break
	} else {
		if needPop := (now.Sub(sample.(*sampleValue).time).Seconds() > s.lifetime.Seconds()); !needPop {
			break
		}
		s.values.Pop()
	}
}
```

File: wavefront/timeUniformSample.go [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+	defer s.mutex.Unlock()
+	return metrics.SampleVariance(s.rawValues())
+}
+
+type sampleValue struct {
+	v    int64
+	time time.Time
+}
+
+func newQueue(size int) Queue {
```
[the remark this answers is not in the document]
Comment (laullon) [reply wavefront/timeUniformSample.go]: take a look of this:
https://github.com/laullon/go_bench_tests

File: wavefront/timeUniformSample.go (hunk 5 above)
Comment (laullon) [reply wavefront/timeUniformSample.go]: sorry, I don't do `break` ;)

File: wavefront/timeUniformSample.go [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+	defer s.mutex.Unlock()
+	s.cleanOldValues()
+	if s.values.IsFull() {
+		s.values.Pop()
+	}
+	s.values.Push(&sampleValue{v: v, time: time.Now()})
+}
+
+func (s *TimeUniformSample) cleanOldValues() {
+	now := time.Now()
```
Comment (venilnoronha) [inline wavefront/timeUniformSample.go]: Need to lock/unlock `s.mutex` here as it's being called from the `Ticker` goroutine.

File: wavefront/timeUniformSample.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (venilnoronha) [reply wavefront/timeUniformSample.go]: This is good. Let's keep the `Queue`.

FYI, we can slightly improve the timing with `slice` based `Queue` by moving to [pointer receivers](https://tour.golang.org/methods/4). See below:

```console
$ go test --bench=. -test.benchmem -cpuprofile profile_cpu.out -memprofile profile_mem.out .
goos: linux
goarch: amd64
pkg: github.com/laullon/go_bench_tests
BenchmarkQueueFix-4     	 5000000	       250 ns/op	      32 B/op	       1 allocs/op
BenchmarkQueueList-4    	 5000000	       292 ns/op	      80 B/op	       2 allocs/op
BenchmarkQueueSlice-4   	 5000000	       246 ns/op	      63 B/op	       1 allocs/op
PASS
ok  	github.com/laullon/go_bench_tests	4.934s
```

[the remark this answers is not in the document]
Comment (laullon) [reply wavefront/timeUniformSample.go]: :O 
I have too much to learn about GO.

File: wavefront/timeUniformSample.go [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return metrics.SampleSum(s.rawValues())
+}
+
+// Update samples a new value.
+func (s *TimeUniformSample) Update(v int64) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	s.cleanOldValues()
```
Comment (venilnoronha) [inline wavefront/timeUniformSample.go]: Once you introduce locking in the `cleanOldValues` method, you'll have to move this call prior to locking within this function as Golang `Mutex` is not reentrant.

File: wavefront/time_uniform_sample.go [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+
+import (
+	"sync"
+	"time"
+
+	metrics "github.com/rcrowley/go-metrics"
+	"istio.io/istio/pkg/log"
+)
+
+// TimeUniformSample is a Sample to store valuse only for a specific time
```
Comment (venilnoronha) [inline wavefront/time_uniform_sample.go]: nit: valuse -> values

File: wavefront/time_uniform_sample.go [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+}
+
+// list of all samples created to be cleaned periodically
+var sampleList = make([]*TimeUniformSample, 0)
+
+// recycle ticker that will fire a samples cleaning
+var ticker *time.Ticker
+
+// time delay for the recycle ticker
+const recycleTime = time.Second * 5
```
Comment (venilnoronha) [inline wavefront/time_uniform_sample.go]: Up to you, but I generally place configurable `const`s right after the `import` statement at the top.

File: wavefront/time_uniform_sample.go [hunk 11]
```diff
@@ -0,0 +1,248 @@
+package wavefront
```
Comment (venilnoronha) [inline wavefront/time_uniform_sample.go]: nit: add license header

File: wavefront/time_uniform_sample_test.go [hunk 12]
```diff
@@ -0,0 +1,73 @@
+package wavefront
```
Comment (venilnoronha) [inline wavefront/time_uniform_sample_test.go]: nit: add license header

File: wavefront/time_uniform_sample.go [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	if s.values.IsFull() {
+		s.values.Pop()
+	}
+	s.values.Push(&sampleValue{v: v, time: time.Now()})
+}
+
+// remove expired values
+func (s *TimeUniformSample) cleanOldValues() {
```
Comment (venilnoronha) [inline wavefront/time_uniform_sample.go]: 2 nits about `var`, `const` and `func` comments:
1. Follow the comment format as described in https://github.com/golang/go/wiki/Comments.
2. Consistently end with a `.` on top level comments.

**Pull request closed without merging**